File: ThreeSum.cc
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

 vector<vector<int> > threeSum(vector<int> &num) {
        vector<vector<int> > results;
        
        if (num.size() < 3)
            return results;

        sort(num.begin(), num.end());
        
        for (int ia = 0; ia < num.size() - 2; ++ia) {
            if ((ia > 0) && (num[ia] == num[ia-1]))
                continue;
                
            int target = -num[ia];
            
            int ib = ia + 1;
            int ic = num.size() - 1;
            
            while (ib < ic) {
                if (num[ib] + num[ic] == target) {
                    vector<int> res{num[ia], num[ib], num[ic]};
                    results.push_back(res);
                    
                    ib++;
                    while ((ib < ic) && (num[ib] == num[ib-1]))
                        ib++;
                    
                    ic--;
                } else if (num[ib] + num[ic] < target) {
                    ib++;
                } else
                    ic--;
            }
            
        }
        
        return results;
    }

int main(int argc, char** argv)
{
    vector<int> test;
    vector<vector<int> > results = threeSum(test);
    for (auto res :results)
    {
        for (auto num: res)
        {
            cout << num << " " ;
        }
        cout << endl;
    }
    return 0;
}

File: ThreeSum.java
public class ThreeSum {

    // O(n^2) time and O(1) space
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> results = new ArrayList<>();
        int target = 0;
        for (int i = 0; i < nums.length; ++i) {
            if ((i < nums.length-1) && (nums[i] == nums[i+1])) {
                continue;
            }
            int start = 0;
            int end = i - 1;
            int other = target - nums[i];
            while (start < end) {
                if ((start > 0) && (nums[start] == nums[start-1])) {
                    start++;
                    continue;
                }
                if ((end < i-1) && (nums[end] == nums[end+1])) {
                    end--;
                    continue;
                }
                int sum = nums[start] + nums[end];
                if (sum == other) {
                    List<Integer> result = new ArrayList<>();
                    result.add(nums[start]);
                    result.add(nums[end]);
                    result.add(nums[i]);
                    results.add(result);
                    end--;
                } else if (sum < other) {
                    start++;
                } else {
                    end--;
                }
            }
        }
        return results;
    }
}